### PR TITLE
Add arbitrary checkpoint capability to `SnapshotContainerBase`

### LINKED ISSUE #27213: Support checkpoint in `SnapshotContainerBase`
### Motivation

Checkpoint is not currently supported in `SnapshotContainerBase` as there is no implementation for storing an loading a `NumericVector<Number> *`.

### Design

Add support for checkpoint in `SnapshotContainerBase` and also add support for storing vectors with arbitrary partitioning, as @maxnezdyur will need this in the future for the flattened jacobian vectors.

### Impact

New capability.

### PR DESCRIPTION
**Pull request by loganharbour**
Closes #27213

### CODE REVIEW

**Comment by loganharbour** [comment]
Still needs a test.

File: framework/src/restart/DataIO.C [hunk 1]
```diff
@@ -611,6 +650,49 @@ dataLoad(std::istream & stream, libMesh::Parameters & p, void * context)
...
+  // if we're advancing back and loading a backup
+  if (!v)
+  {
+    v = NumericVector<Number>::build(comm, solver_package);
+    v->init(size, local_size);
+  }
+  else
+    mooseAssert(v->type() != GHOSTED, "Cannot be ghosted");
+
+  // Make sure that the sizes are consistent; this will happen if we we're calling this
```
Comment (grmnptr) [inline framework/src/restart/DataIO.C:687]:
```suggestion
  // Make sure that the sizes are consistent; this will happen if we're calling this
```

**Review by grmnptr — APPROVED** [review]
Looks good to me, thanks for adding this capability!

**Review by grmnptr — CHANGES_REQUESTED** [review]
Ups

**Review by grmnptr — APPROVED** [review]
Nice!

**Review by grmnptr — APPROVED** [review]

**Pull request merged**